### ISSUE TITLE
🛠 Tooling: Find a way to avoid 'as unknown as ESTree.Node' assertions

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [JoshuaKGoldberg/eslint-plugin-package-json#125]
### Bug Report Checklist

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have pulled the latest `main` branch of the repository.
- [X] I have [searched for related issues](https://github.com/JoshuaKGoldberg/eslint-plugin-package-json/issues?q=is%3Aissue) and found none that matched my issue.

### Overview

Following #40 & #59 & https://github.com/ota-meshi/jsonc-eslint-parser/issues/184: the current state of types for AST nodes in this plugin is that we've completely swapped out the built-in ESLint ones. Nodes are instead typed using `import type { AST as JsonAST } from "jsonc-eslint-parser";` -> `node: JsonAST.JSONProperty` and the like.

Unfortunately, ESLint's APIs still expect ESTree nodes. So we end up with assertions any time we need to pass the nodes to ESLint's APIs:

https://github.com/JoshuaKGoldberg/eslint-plugin-package-json/blob/84035c3f2bac376567eef767ba21f354a0d93f54/src/rules/valid-repository-directory.ts#L32-L34

This isn't ideal. One workaround done by `eslint-plugin-jsonc` is to [instead pass `loc: node.loc`](https://github.com/ota-meshi/eslint-plugin-jsonc/blob/bb0cc408c928ef12952d140add87f3c2b8a2eb40/lib/rules/no-bigint-literals.ts#L26-L27). But I'd ideally like to be able to just pass the `node` directly.

### Additional Info

I don't know a straightforward way to make this better. `context.report` is typed in `@types/eslint` as:

```ts
report(descriptor: ReportDescriptor): void;
```

...with:

```ts
type ReportDescriptor = ReportDescriptorMessage & ReportDescriptorLocation & ReportDescriptorOptions;
type ReportDescriptorMessage = { message: string } | { messageId: string };
type ReportDescriptorLocation =
    | { node: ESTree.Node }
    | { loc: AST.SourceLocation | { line: number; column: number } };
```

I suppose `@types/eslint` could mark the context with a type parameter like `<Node = ESTree.Node>` 😬... Eww.

**Comment by michaelfaith** [comment]
Curious if it makes sense to still pursue this if the ultimate direction is to move to `@eslint/json` (https://github.com/JoshuaKGoldberg/eslint-plugin-package-json/issues/655)?

**Comment by JoshuaKGoldberg** [comment]
It'd be great if resolving #655 resolved this too, yes!
